sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -161,22 +164,45 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
           bucketSpec,
           Some(partitionSchema))
 
+        val catalogTable = metastoreRelation.catalogTable
         val logicalRelation = cached.getOrElse {
           val sizeInBytes =
             metastoreRelation.stats(sparkSession.sessionState.conf).sizeInBytes.toLong
           val fileIndex = {
-            val index = new CatalogFileIndex(
-              sparkSession, metastoreRelation.catalogTable, sizeInBytes)
+            val index = new CatalogFileIndex(sparkSession, catalogTable, sizeInBytes)
             if (lazyPruningEnabled) {
               index
             } else {
               index.filterPartitions(Nil)  // materialize all the partitions in memory
             }
           }
           val partitionSchemaColumnNames = partitionSchema.map(_.name.toLowerCase).toSet
-          val dataSchema =
-            StructType(metastoreSchema
-              .filterNot(field => partitionSchemaColumnNames.contains(field.name.toLowerCase)))
+          val filteredMetastoreSchema = StructType(metastoreSchema
+            .filterNot(field => partitionSchemaColumnNames.contains(field.name.toLowerCase)))
+
+          // Infer a case-sensitive schema when the metastore doesn't return one, if configured.
+          val inferredSchema = inferSchema(
+            catalogTable,
+            metastoreSchema,
+            options,
+            defaultSource,
+            fileType,
+            fileIndex)
+
+          // If configured, save the inferred case-sensitive schema to the table properties and
+          // fetch the updated CatalogTable record for use in the LogicalRelation.
+          val updatedCatalogTable = updateCatalogTable(catalogTable, inferredSchema)
+
+          val dataSchema = inferenceMode match {
+            case (INFER_AND_SAVE | INFER_ONLY) if (!catalogTable.schemaPreservesCase) =>
+              inferredSchema.getOrElse {
+                logWarning(s"Unable to infer schema for table $tableIdentifier from file format " +
+                  s"$defaultSource (inference mode: $inferenceMode); using metastore schema.")
+                filteredMetastoreSchema
+              }
+            case _ =>
+              filteredMetastoreSchema
+          }
```

Comment:
I'll see if I can clean this up some more. It gets a bit messy since we'll want to replace ```catalogTable``` in the ```LogicalRelation``` returned by this method with the updated catalog table returned by ```updateTableCatalog()``` if the schema is successfully written to the metastore. Essentially, we need to return an ```Option[StructType]``` representing if the schema was inferred or not and an ```Option[CatalogTable]``` representing if the catalog table was updated.